src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -263,57 +266,98 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
-            var knownProjects = projects.Select(e => e.MSBuildProjectPath).ToHashSet(PathUtility.GetStringComparerBasedOnOS());
+            var knownProjects = new ConcurrentDictionary<string, bool>(PathUtility.GetStringComparerBasedOnOS());
+            knownProjects.AddRange(projects.Select(e => e.MSBuildProjectPath)
+                .Select(proj => new KeyValuePair<string, bool>(proj, false)));
 
-            for (var i = 0; i < projects.Count; i++)
+            var options = new ExecutionDataflowBlockOptions()
             {
-                var (packageSpecs, projectAdditionalMessages) = await projects[i].GetPackageSpecsAndAdditionalMessagesAsync(context);
+                MaxDegreeOfParallelism = Environment.ProcessorCount
+            };
+            var actionBlock = new ActionBlock<ProjectRestoreSpec>(GetProjectRestoreSpecAndAdditionalMessages, options);
 
-                if (projectAdditionalMessages != null && projectAdditionalMessages.Count > 0)
+            foreach (IDependencyGraphProject project in projects)
+            {
+                var (packageSpecs, projectAdditionalMessages) = await project.GetPackageSpecsAndAdditionalMessagesAsync(context);
+                await actionBlock.SendAsync(new ProjectRestoreSpec()
                 {
-                    if (allAdditionalMessages == null)
-                    {
-                        allAdditionalMessages = new List<IAssetsLogMessage>();
-                    }
+                    DgSpec =  dgSpec,
+                    PackageSpecs = packageSpecs,
+                    ProjectAdditionalMessages = projectAdditionalMessages,
+                    KnownProjects = knownProjects,
+                    AllAdditionalMessages = allAdditionalMessages
+                });
+            }
+
+            actionBlock.Complete();
+            await actionBlock.Completion;
+
+            // Return dg file
+            return (dgSpec, allAdditionalMessages?.ToList());
+        }
+
+        private static void GetProjectRestoreSpecAndAdditionalMessages(
+            ProjectRestoreSpec restoreSpecData)
+        {
+            var dgSpec = restoreSpecData.DgSpec;
+
+            if (restoreSpecData.ProjectAdditionalMessages?.Any() ?? false)
+            {
+                if (restoreSpecData.AllAdditionalMessages == null)
+                {
+                    restoreSpecData.AllAdditionalMessages = new ConcurrentBag<IAssetsLogMessage>();
+                }
 
-                    allAdditionalMessages.AddRange(projectAdditionalMessages);
+                foreach (var projectAdditionalMessage in restoreSpecData.ProjectAdditionalMessages)
+                {
+                    restoreSpecData.AllAdditionalMessages.Add(projectAdditionalMessage);
                 }
+            }
 
-                foreach (var packageSpec in packageSpecs)
+            foreach (var packageSpec in restoreSpecData.PackageSpecs)
+            {
+                lock (dgSpec)
```

Comment:
It would be nice to have the collections on the DGSpec be concurrent and avoid the lock.  I am not sure if it is going to be fixed but could we add a tracking issue for this?

Reply:
We had discussion on this [Issue#9002](https://github.com/NuGet/Home/issues/9002) and after [some inputs](https://github.com/NuGet/NuGet.Client/pull/3594#issuecomment-682199570) we decided not to implement it.